Step3 PR

yunbaek opened the pull request:
안녕하세요! 백경윤입니다.

3단계 미션 PR 입니다!

yunbaek committed:
refactor: 피드백 반영

yunbaek committed:
docs: step3-로또 요구사항 정리

yunbaek committed:
feat: 로또 번호 검증 기능 추가

Commit:
refactor: static factory method 사용

Commit:
feat: 로또 티켓 생성 기능 추가

Commit:
refactor: 리팩토링

- Money 객체 추출
- 상수 추출
- 번호 생성 전략 패턴 적용

Commit:
refactor: 로또 번호 생성 팩토리 메서드 수정

- argument: Set<LottoNumbers> -> Set<Integer>

Commit:
feat: 사용자 입력 기능 추가

yunbaek committed:
chore: package naming 변경

- util -> converter

yunbaek committed:
refactor: LottoTicket 일급 컬렉션 추가

yunbaek committed:
feat: 로또 개수, 숫자 출력 기능 추가

yunbaek committed:
feat: 당첨 티켓 생성 로직 추가

yunbaek committed:
feat: 순위 기능 구현

yunbaek committed:
refactor: 금액 data type - int -> long

yunbaek committed:
feat: 수익률 계산 로직 추가

yunbaek committed:
feat: 결과 순서 정렬 기능 구현

yunbaek committed:
feat: 출력용 결과 필터, 초기화 적용

yunbaek committed:
feat: 게임 실행 구현

devyonghee commented on README.md:
```diff
@@ -38,4 +38,44 @@
+  - 4개 일치 (50_000원) 
+  - 5개 일치 (1_500_000원) 
+  - 6개 일치 (2_000_000_000원)
+  - 총 구매 금액 대비 당첨 금액의 수익률을 계산하여 출력
```
기능별로 요구사항을 정리해주셨군요👍👍👍
그런데 요구사항이 실수로 두번 들어간 것 같아요 😅

devyonghee commented on src/main/java/lotto/controller/LottoGameController.java:
```diff
@@ -0,0 +1,53 @@
+import lotto.domain.strategy.RandomGenerateStrategy;
+import lotto.view.InputView;
+import lotto.view.ResultView;
+
+public class LottoGameController {
+
```
mvc 패턴으로 구분해서 잘 구현해주신 것 같아요 😆👍

devyonghee commented on src/main/java/lotto/controller/LottoGameController.java:
```diff
@@ -0,0 +1,53 @@
+        resultView.printLottoTickets(lottoTickets);
+        return lottoTickets;
+    }
+
+    private WinningLottoTicket winningTicket() {
+        LottoNumbers lottoNumbers = inputView.readWinningNumbers();
+        return WinningLottoTicket.of(lottoNumbers);
+    }
+
+    private LottoTickets getLottoTickets(Money money) {
+        LottoTicketMachine lottoTicketMachine = new LottoTicketMachine(new RandomGenerateStrategy());
+        return lottoTicketMachine.buyLottoTickets(money);
+    }
```
이전 단계에서 코멘트 해드려서 아시겠지만
메소드명에 대해서는 하나의 컨벤션을 지키는 것을 추천드릴게요!! 😄
팀원과 협업하는 경우라면 혼동이 생길 수 있을 것 같아서요!

yunbaek replied:
피드백 주신대로 빌더와 조정자를 나눠 네이밍을 하려고 해봤는데, 아직 조금은 어색하네요..
그래서 실수 한 것 같습니다 😂 
다시 한 번 살펴보겠습니다!

devyonghee commented on src/main/java/lotto/controller/LottoGameController.java:
```diff
@@ -0,0 +1,53 @@
+        this(new InputView(), new ResultView());
+    }
+
+    public void play() {
+        Money money = inputView.readMoney();
+        LottoTickets lottoTickets = purchasedTickets(money);
+        WinningLottoTicket winningLottoTicket = winningTicket();
+        Result result = Result.from(lottoTickets, winningLottoTicket);
+        resultView.printResult(result);
+        Money totalPrize = Money.of(result.totalPrize());
+        resultView.printProfitRate(totalPrize.profitRate(money));
+    }
```
개인적으로 모든 행위에 대해서 로컬 변수를 선언이 필요한가 라는 의문이 생기는데요 🤔
참조가 바로 제거되어 해당 변수는 마이너 gc 에서 제거되긴 하겠지만 가독성에 어려움이 생기는 것 같아서요.
가독성을 위해 역할별로 라인을 구분해주거나 적절하게 인라인을 활용해보는 방법은 어떻게 생각하시나요?

yunbaek replied:
말씀주신대로 가독성이 많이 떨어지는 것 같습니다.
라인을 구분해서 가독성을 높여보겠습니다!

devyonghee commented on src/main/java/lotto/converter/InputConverter.java:
```diff
@@ -0,0 +1,57 @@
+		int integer;
+		try {
+			integer = Integer.parseInt(input);
+		} catch (NumberFormatException e) {
+			throw new InvalidUserInputException("숫자를 입력해주세요.");
+		}
```
이전 단계에서 코멘트를 해드리긴 했지만 제 설명이 부족했던 것 같아 추가 코멘트를 드리겠습니다. 😥
어떤 값이 들어와서 잘못되었는지 더 자세한 내용을 추가해주신 다면 예외가 발생했을 때 디버깅이 수월해질 것 같은데 이러한 방향에 대해 어떻게 생각하시나요??
제가 참고했던 링크 남겨드리겠습니다!! [예외의 상세 메시지에 실패 관련 정보를 담으라](https://codingwell.tistory.com/157?category=1013497)

yunbaek replied:
아 제가 피드백 주셨던 내용을 제대로 이해하지 못했었네요ㅠㅠ
네! 피드백 처럼 예외가 발생한 이유를 좀 더 명확히 전달해주겠습니다!!

devyonghee commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,46 @@
+        this.value = value;
+    }
+
+    public static LottoNumber of(int value) {
+        return new LottoNumber(value);
+    }
```
정적 팩토리 메소드를 이용해서 객체 생성을 해주셨군요 👍
아시겠지만 이펙티브 자바에서도 정적 팩토리 메서드 사용을 권장하고 있는데요.
이 책에서 명명 규칙을 소개하고 있는데 인자가 하나인 경우에는 `from` 을 얘기하고 있습니다. 
수정이 필요한 건 아니지만 참고하시면 좋을 것 같아 링크로 남겨드리겠습니다
[생성자 대신 정적 팩토리 메서드를 고려하라](https://sun-22.tistory.com/84)

yunbaek replied:
에고.. 분명히 본 기억이 있었는데 `from` 과 `of` 를 햇갈렸네요 😂 
네이밍도 신경쓰겠습니다!

devyonghee commented on src/main/java/lotto/domain/LottoNumbers.java:
```diff
@@ -0,0 +1,67 @@
+		return Objects.hash(lottoNumbers);
+	}
+
+	public int compare(LottoNumbers purchaseLottoNumbers) {
+		return (int) this.lottoNumbers.stream()
+			.filter(purchaseLottoNumbers.lottoNumbers::contains)
+			.count();
+	}
```
`compare` 란 메소드 명은 `Comparator` 의 추상 메소드로  많은 곳에서 두 객체의 크기를 비교하여 그 결과로 `-1`, `0`, `1` 을 반환하고 있어요. (ex. `Integer::compare`, `Long::compare`)
많은 곳에서 이렇게 사용되고 있다보니 메소드명을 보면 클라이언트에게 혼동을 줄 수 있을 것 같아요 🥲 
혹시 다른 메소드 명을 고려해보시는 건 어떻게 생각하시나요??🤔

yunbaek replied:
미처 신경쓰지 못한 부분이었습니다!!
`compare` ->  `matchCount` 정도로 바꿔보겠습니다!

devyonghee commented on src/main/java/lotto/domain/LottoTicket.java:
```diff
@@ -0,0 +1,50 @@
+import lotto.domain.strategy.GenerateStrategy;
+
+public class LottoTicket {
+	private static final String OPEN_BRACKET = "[";
+	private static final String CLOSE_BRACKET = "]";
```
개인적으로 비즈니스 의미를 가진 매직 리터럴의 경우에만 상수로 선언해서 사용하고 있는데요. 
`toString` 을 위한 괄호에도 비즈니스 의미를 가지고 있는 것인지는 고민되는 부분인 것 같습니다 🤔🤔
경윤님도 같이 고민해보시면 좋을 것 같아 코멘트 남겨드립니다!!
제가 참고한 내용에 대해 링크 남겨드리도록 하겠습니다!! [매직넘버, 매직 리터럴](https://limdingdong.tistory.com/5)

yunbaek replied:
그렇지 않아도 `toString` 을 위해 상수 추출을 하는 것에 대해 고민했는데 비즈니스적 의미를 띄고 있지 않아 굳이 필요없다는 데에 저도 동의합니다!
다만 `toString` 이 비즈니스적으로 중요할 때에는 상수 추출을 적극 검토해보겠습니다ㅏ!

devyonghee commented on src/main/java/lotto/domain/LottoTicketMachine.java:
```diff
@@ -0,0 +1,38 @@
+
+public class LottoTicketMachine {
+
+	private static final int LOTTO_COST_PER_TICKET = 1000;
+	private static final int START_INCLUSIVE = 0;
+	private final GenerateStrategy generateStrategy;
+	private final Money ticketPrice;
+
+	public LottoTicketMachine(GenerateStrategy generateStrategy) {
+		this.generateStrategy = generateStrategy;
+		this.ticketPrice = Money.of(LOTTO_COST_PER_TICKET);
+	}
```
로또 금액은 원시 타입 상수로 가지고 있고 객체가 생성될 때 `Money` 로 변환해서 사용하게 되는 것 같아요.
혹시 미리 `Money` 를 상수로 가지고 있는 것이 아니라 이렇게 구현하신 이유가 있으실까요?? 🤔

로또 금액을 상수로 가지고 있다면 
`LottoTicketMachine` 이 생성될 때마다 `ticketPrice` 을 생성하는 것이 아니라 재활용이 가능해질 것 같아서요!

yunbaek replied:
처음에 금액을 `int` 로 구현했다가 wrapping 하는 과정에서 큰 고민 없이 `Money` 객체가 필요하니 생성자에서 생성하도록 구현했습니다 😂 
말씀하신대로 티켓 장당 가격을 상수로 선언하고 재사용 하는 것이 맞다고 생각됩니다!

devyonghee commented on src/main/java/lotto/domain/LottoTicketMachine.java:
```diff
@@ -0,0 +1,38 @@
+	private LottoTickets lottoTickets(TicketCount ticketCount) {
+		return IntStream.range(START_INCLUSIVE, ticketCount.count())
+			.mapToObj(i -> this.generateLottoTicket(generateStrategy))
+			.collect(Collectors.collectingAndThen(Collectors.toList(), LottoTickets::of));
```
`Collector` 를 잘 활용해주신 것 같아요!! 👍
이렇게 사용하는 건 본적이 없었는데 괜찮은 방법 같아요 😆

devyonghee commented on src/main/java/lotto/domain/LottoTickets.java:
```diff
@@ -0,0 +1,42 @@
+
+    private LottoTickets(List<LottoTicket> lottoTickets) {
+        this.lottoTickets = lottoTickets;
+    }
```
내부의 데이터를 보호하기 위해서 컬렉션의 주소값 공유를 외부와 끊어주는 것이 좋다고 생각하는데요.
방어적 프로그래밍을 위해 값을 할당할 때 새로운 리스트로 복사해주는 방법은 어떻게 생각하시나요??
제가 참고했던 자료에 대해서 남겨드리도록 하겠습니다!! [방어적 복사란](https://velog.io/@miot2j/%EC%96%95%EC%9D%80%EB%B3%B5%EC%82%AC-%EA%B9%8A%EC%9D%80%EB%B3%B5%EC%82%AC-%EB%B0%A9%EC%96%B4%EC%A0%81-%EB%B3%B5%EC%82%AC%EB%9E%80)

yunbaek replied:
이 부분은 사실 그 동안 코딩하면서 크게 신경 쓰지 않았던 부분입니다.
덕분에 `불변`에 대해 조금 더 알아보게 되는 계기가 되었습니다.

지금은 일단 `LottoTickets` 클래스만 수정하였는데,
앞으로 피드백 주신 내용 기억하고, 적용해보도록 하겠습니다!!

devyonghee commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,67 @@
+		return value;
+	}
+
+	public TicketCount ticketCount(Money ticketPrice) {
+		if (this.value < PRICE_PER_TICKET) {
+			throw new InvalidMoneyException(INVALID_INPUT_MONEY_MESSAGE);
+		}
+		try {
+			return TicketCount.of((int) (this.value / ticketPrice.getValue()));
+		} catch (ArithmeticException e) {
+			throw new InvalidMoneyException(INVALID_TICKET_PRICE_MESSAGE);
+		}
+	}
```
직접적으로 import 를 하지 않았지만 메소드 명, 파라미터 명, 반환 타입들로 `Money` 에서 ticket 에 대한 의존이 간접적으로 생기게 된 것 같아요.
의존성은 한 방향으로 향해야 한다고 생각하여 `Money` 에서 ticket 을 모르게 하면 좋을 것 같은데 이 방향에 대해 어떻게 생각하시나요??
제가 고민했던 자료 남겨드리도록 하겠습니다!! [조영호작가님의 우아한 객체지향 강의 후기 및 정리](https://ramees.tistory.com/47)

yunbaek replied:
네! 다시 살펴보니 `Money` 객체는 일종의 VO 객체인데, 이곳저곳에서 쓰이다보니 불필요한 의존성이 생긴 것 같습니다.
말씀대로 `Money` 라는 객체는 `ticket` 에 관련한 정보들을 몰라도 되는 것 같습니다.

`Money` 에서 `티켓 수` 를 구하는게 아니라, 값을 나눠주는 `divide` 메소드만 제공하고 `LottoTicketMachine` 에서 `티켓 수` 를 구하는 것이 바람직할 것 같아서 이렇게 수정해보겠습니다! :smile:

devyonghee commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.util.Arrays;
+
+public enum Rank {
+	FIRST(6, 2_000_000_000),
+	SECOND(5, 1_500_000),
+	THIRD(4, 50_000),
+	FOURTH(3, 5_000),
+	LOSE(0, 0);
```
순위를 `Enum` 으로 관리 해주신 모습이 깔끔하고 아주 좋은 것 같아요 👍👍👍

devyonghee commented on src/main/java/lotto/domain/Result.java:
```diff
@@ -0,0 +1,36 @@
+			.sum();
+	}
+
+	public String toString() {
+		return rankResult.entrySet().stream()
+			.map(entry -> entry.getKey().getMatchCount() + "개 일치 (" + entry.getKey().getPrize() + "원) - " + entry.getValue() + "개")
+			.collect(Collectors.joining(System.lineSeparator()));
+	}
```
개인적으로 도메인의 영역은 코드 유지 보수를 위해 순수 자바 코드가 될수록 좋다고 생각하는데요.
그런데 이 부분에서는 `System.lineSeparator` 을 포함하여 뷰의 영역까지 고려하게 되는 것 같아요 😅
혹시 이러한 의존 방향에 대해서 어떻게 생각하시나요??🤔

yunbaek replied:
말씀하신 대로 도메인이 너무 뷰 영역까지 신경쓰고 있지 않나 고민을 하긴 했었는데.. 

1. `View layer` 에서의 사용을 위해 일급 컬렉션으로 감싸놓은 `Map<Rank, Long> rankResult` 객체를 굳이 `getter` 를 선언하여 다시 꺼내 쓰는게 옳을까? 
2. `Results` 객체의 `toString` 을 기존 코드대로 재정의 하면, `View layer` 에서 편하게 출력할 수 있을 것 같아..

라는 생각으로 저런식으로 구현했습니다.
특히 2번은 당장의 편함을 택한 건데 바로 집어내셔서 창피하네요 😂 

피드백 주신 것 처럼 도메인이 뷰의 책임까지 갖고 있는 건 바람직하지 않은 것 같아, 분리하도록 하겠습니다! 😄

devyonghee commented on src/main/java/lotto/domain/strategy/GenerateStrategy.java:
```diff
@@ -0,0 +1,7 @@
+
+import lotto.domain.LottoNumbers;
+
+public interface GenerateStrategy {
+	LottoNumbers generate();
+}
```
전략 패턴을 이용하여 확장성을 고려해주신 점 아주 인상적입니다!!!! 👍👍👍👍👍

devyonghee commented on src/main/java/lotto/domain/strategy/RandomGenerateStrategy.java:
```diff
@@ -0,0 +1,22 @@
+
+public class RandomGenerateStrategy implements GenerateStrategy {
+	private static final Random RANDOM = new Random();
+	private static final int MAX_LOTTO_NUMBER_SIZE = 6;
+	private static final int MINIMUM_NUMBER_BOUND = 1;
+	private static final int MAXIMUM_NUMBER_BOUND = 45;
```
`LottoNumber`, `LottoNumbers` 에서 선언된 값들과 중복되는 것 같아요. 
상수를 내부로 감춘 것은 좋긴 하지만 오히려 응집도가 떨어져 관리 해야 할 코드가 증가된 것 같은데요. 
이러한 정보들을 하나의 객체로 캡슐화하거나 패키지 내에서만 사용할 수 있도록 상수를 열어주는 방향에 대해서는 어떻게 생각하시나요??🤔

yunbaek replied:
중복되는 사용을 체크하지 못했네요 😢 
말씀대로 비즈니스적으로 중요한 값들인데 응집도가 너무 떨어지는 것 같습니다.
비즈니스적으로 중요하다 생각되어 상수를 열어주는 것 보다 `enum` 을 활용하여 리팩토링 해 보겠습니다! 😄

devyonghee commented on src/main/java/stringaddcalculator/StringAddCalculator.java:
```diff
@@ -27,23 +31,19 @@ public static int splitAndSum(String input) {
 		if (number < ZERO_NUMBER) {
-			throw new NegativeNumberException();
+			throw new NegativeNumberException(NEGATIVE_NUMBER_EXCEPTION_MESSAGE);
 		}
```
피드백을 적극 수용해주시고 반영해주셔서 감사합니다 🙇‍♂️🙇‍♂️🙇‍♂️🙇‍♂️

devyonghee commented on src/main/java/lotto/exception/InvalidLottoNumberException.java:
```diff
@@ -0,0 +1,7 @@
+package lotto.exception;
+
+public class InvalidLottoNumberException extends RuntimeException {
+    public InvalidLottoNumberException(String message) {
+        super(message);
+    }
+}
```
커스텀 에외를 선언해서 사용해주신 점 좋은 것 같아요 👍👍

devyonghee commented on src/test/java/lotto/domain/RanksTest.java:
```diff
@@ -0,0 +1,39 @@
+
+	@Test
+	@DisplayName("FOURTH 부터 역순으로 정렬 된 순위 별 개수 map 반환")
+	void groupByTest() {
+		Ranks ranks = Ranks.of(List.of(6, 5, 4, 6, 6));
+		Map<Rank, Long> groupBy = ranks.groupBy();
+		assertAll(
+			() -> assertThat(groupBy).containsEntry(Rank.FOURTH, 0L),
+			() -> assertThat(groupBy).containsEntry(Rank.THIRD, 1L),
+			() -> assertThat(groupBy).containsEntry(Rank.SECOND, 1L),
+			() -> assertThat(groupBy).containsEntry(Rank.FIRST, 3L),
+			() -> assertThat(groupBy.keySet().toArray()[0]).isEqualTo(Rank.FOURTH),
+			() -> assertThat(groupBy.keySet().toArray()[1]).isEqualTo(Rank.THIRD),
+			() -> assertThat(groupBy.keySet().toArray()[2]).isEqualTo(Rank.SECOND),
+			() -> assertThat(groupBy.keySet().toArray()[3]).isEqualTo(Rank.FIRST)
+		);
+	}
```
`containsExactlyEntriesOf` 을 이용하시면 내부에서 순서도 같이 검증해주기 때문에
반복되는 검증 메서드 호출 없이 원하시는 검증을 그대로 수행할 수 있을 것 같아요!!
이렇게 검증할 수도 있으니 참고만 해주세요!! 😁
```suggestion
	void groupByTest() {
		//given
		Ranks ranks = Ranks.of(List.of(6, 5, 4, 6, 6));
		LinkedHashMap<Rank, Long> expected = new LinkedHashMap<>();
		expected.put(Rank.FOURTH, 0L);
		expected.put(Rank.THIRD, 1L);
		expected.put(Rank.SECOND, 1L);
		expected.put(Rank.FIRST, 3L);
		//when
		Map<Rank, Long> groupBy = ranks.groupBy();
		//then
		assertThat(groupBy).containsExactlyEntriesOf(expected);
	}
```

yunbaek replied:
저런 메서드가 있는 지 몰랐네요 😂 
그리고 테스트 메서드가 조금 길어졌을 때, 
`given, when, then` 으로 나누니 훨씬 가독성이 좋아지네요!!
좋은 조언 감사드립니다!

devyonghee reviewed (changes requested):
안녕하세요 경윤님!
로또 (자동) 잘 구현해주신 것 같아요 😆👍👏
요구사항도 잘 정리해주시고 특히 전략 패턴을 이용하여 확장성을 고려해주신 점, mvc 패턴으로 구현해주신 점 아주 인상 깊었습니다!!

추가적으로 고민해보시면 좋은 부분들에 대해 코멘트 남겨드렸는데
이에 대한 의견 부탁드릴게요!! 🙏🙇‍♂️🙇‍♂️

혹시 궁금하시거나 이상한 점이 보인다면 편하게 DM 이나 코멘트 남겨주시기 바랍니다. 😆

추가적으로 혹시 IDE 에서 Java 9 버전 이상을 사용하고 계신지 확인 한번만 부탁드리겠습니다.
기본적으로 build.gradle 설정에는 Java 8 버전으로 설정되어 있어서 문제가 있을 수 있어서요!!

Commit:
docs: 요구사항 중복 수정

Commit:
refactor: 피드백 반영

- 인라인, 줄바꿈 활용하여 가독성 재고
- 에러 메시지 내용 추가
- 테스트 메서드 수정
- 정적 팩토리 메서드 네이밍 수정
- `compare` 메서드 네이밍 변경
- 티켓 장당 가격 `Money` 객체 상수 추출
- 로또 정보 enum 사용

Commit:
refactor: Result 리팩토링

- 출력 책임 이동(domain -> view)
- 수익률 반환 책임 이동(money -> result)

Commit:
fix: 오타 수정

Commit:
refactor: ticket 수 관련 로직 위치 변경

- Money -> LottoTicketMachine

Commit:
feat: 에러메시지 처리 로직 추가

Commit:
refactor: 방어적 복사 적용

devyonghee commented on src/main/java/lotto/domain/LottoInfo.java:
```diff
@@ -0,0 +1,19 @@
+package lotto.domain;
+
+public enum LottoInfo {
+
+	MAX_LOTTO_NUMBER(45),
+	MIN_LOTTO_NUMBER(1),
+	MAX_LOTTO_NUMBER_SIZE(6),
+	;
```
음...상수로 선언해서 사용하는 모습과 비슷한 구조 같아보이는데요. 😅
오히려 잘못된 추상화로 인해 확장성에 방해가 되지 않을까 우려됩니다.
우선 외부로 노출되지 않기 위해 `enum` 의 접근 지정자를 `default` 로 수정해보시는 것은 어떨까 생각하는데 어떠신가요? 🤔

devyonghee commented on src/main/java/lotto/domain/LottoTicketMachine.java:
```diff
@@ -0,0 +1,42 @@
+	}
+
+	private TicketCount getTicketCount(Money money) {
+		if (money.getValue() < LOTTO_COST_PER_TICKET) {
+			throw new InvalidMoneyException(INVALID_INPUT_MONEY_MESSAGE);
+		}
```
디미터 법칙을 지키기 위해 getter 로 내부 값을 꺼내와서 비교하는 것보다 
`money` 에게 메소드를 추가하는 방식을 추천드리겠습니다! 😆

```suggestion
		if (money.isLessThan(PRICE_PER_TICKET)) {
			throw new InvalidMoneyException(INVALID_INPUT_MONEY_MESSAGE);
		}
```

devyonghee commented on src/main/java/lotto/domain/Ranks.java:
```diff
@@ -0,0 +1,38 @@
+				.map(Rank::from)
+				.collect(Collectors.toList()));
+	}
+	private Map<Rank, Long> initResultMap() {
+		Map<Rank, Long> resultRankMap = new LinkedHashMap<>();
+		Arrays.stream(Rank.values())
+			.filter(r -> !r.equals(Rank.LOSE))
+			.sorted(Comparator.comparingInt(Rank::getMatchCount))
+			.forEach(r -> resultRankMap.put(r, 0L));
+		return resultRankMap;
+	}
+
+
+	public Map<Rank, Long> groupBy() {
+		Map<Rank, Long> resultRankMap = initResultMap();
+		ranks.forEach(rank -> resultRankMap.computeIfPresent(rank, (key, value) -> ++value));
+		return resultRankMap;
+	}
```
반드시 수정해야할 부분은 아니지만 참고하시면 좋을 것 같아 코멘트 남겨드립니다.

컨벤션에 따라 달라지긴 하겠지만 개인적으로 코드는 책처럼 읽어 내려가면 좋다고 생각하는데요.
아래에 있는 메소드에서 위의 메소드를 호출하고 있기 때문에 두 메소드의 위치를 변경하는 것은 어떤가요??

제가 참고하고 고민했던 클린 코드 책의 내용 남겨드리도록 하겠습니다! [위에서 아래로 코드 읽기: 내려가기 규칙](https://holika.tistory.com/59)

devyonghee commented on src/main/java/lotto/domain/Result.java:
```diff
@@ -0,0 +1,44 @@
+		return new Result(rankResult, profitRate(rankResult, inputMoney));
+	}
+
+	public static Result of(LottoTickets purchasedTickets, WinningLottoTicket winningTicket, Money inputMoney) {
+		Ranks ranks = Ranks.from(purchasedTickets.match(winningTicket));
+		Map<Rank, Long> resultMap = ranks.groupBy();
+		double profitRate = profitRate(resultMap, inputMoney);
+		return new Result(resultMap, profitRate);
+	}
```
이 정적 팩토리 메소드에서 매칭 개수를 세기, 랭크들을 생성, 랭크 그룹 형태 조회, 수익률 계산을 수행하고 있는데요.
이러한 역할들이 정적 팩토리 메소드에 위치하고 있는 것이 적절한지 고민해보시면 좋을 것 같아요 😄

devyonghee commented on src/main/java/lotto/domain/WinningLottoTicket.java:
```diff
@@ -0,0 +1,17 @@
+		return new WinningLottoTicket(lottoNumbers);
+	}
+
+	public int matchCount(LottoTicket lottoTicket) {
+		LottoNumbers winningNumbers = super.lottonumbers();
```
부모 클래스는 데이터를 저장하는 용도로만 사용하고 자식 클래스에서는 메소드로 다시 꺼내서 사용하고 있는데요. 

부모 클래스 메소드를 활용하여 중복 코드를 제거할 순 있지만 
이런 구조는 캡슐화가 깨지기 쉬우며, 결합도가 높아지게 되어 나중에는 부모 클래스 수정이 어려워질 것 같아요.
 
그래서 저는 상속은 템플릿 메소드 패턴 이외에는 선호하지 않는 편이긴 합니다 😄
경윤님도 같이 고민해보시면 좋을 것 같아 코멘트 남겨드립니다!!😆

devyonghee reviewed (approved):
안녕하세요 경윤님
피드백을 수용하시고 반영해주셔서 진심으로 감사합니다 🙇‍♂️🙇‍♂️

깔끔한 코드를 위해 많이 고민해주신 것 같아 감동입니다 🥹
추가적으로 참고해보시면 좋은 것 같은 부분 코멘트 남겨드렸는데
확인 한번만 부탁드리겠습니다. 🙏

이번 미션은 여기서 마무리 하고 머지하도록 하겠습니다!
피드백 반영해주시느라 정말 고생 많으셨고 다음 로또 2등 미션도 파이팅입니다 💪🚀
